VultisigApp/VultisigApp/Services/CoinService.swift

```diff
@@ -0,0 +1,106 @@
+//
+//  CoinService.swift
+//  VultisigApp
+//
+//  Created by Enrique Souza Soares on 05/07/24.
+//
+
+import Foundation
+
+@MainActor
+class CoinService {
+    
+    static let shared = CoinService()
+    
+    public func removeCoins(coins: [Coin], vault: Vault) async throws {
+        for coin in coins {
+            if let idx = vault.coins.firstIndex(where: { $0.ticker == coin.ticker && $0.chain == coin.chain }) {
+                vault.coins.remove(at: idx)
+            }
+            
+            await Storage.shared.delete(coin)
+        }
+    }
+    
+    public func saveAssets(for vault: Vault, selection: Set<CoinMeta>) async {
+        do {
+            let removedCoins = vault.coins.filter { coin in
+                !selection.contains(where: { $0.ticker == coin.ticker && $0.chain == coin.chain})
+            }
+            let nativeCoins = removedCoins.filter { $0.isNativeToken }
+            let allTokens = vault.coins.filter { coin in
+                nativeCoins.contains(where: { $0.chain == coin.chain }) && !coin.isNativeToken
+            }
+            
+            try await removeCoins(coins: removedCoins, vault: vault)
+            try await removeCoins(coins: nativeCoins, vault: vault)
+            try await removeCoins(coins: allTokens, vault: vault)
+            
+            // remove all native tokens and also the tokens so they are not added again
+            let filteredSelection = selection.filter{ selection in
+                !nativeCoins.contains(where: { selection.ticker == $0.ticker && selection.chain == $0.chain}) &&
+                !allTokens.contains(where: { selection.ticker == $0.ticker && selection.chain == $0.chain})
+            }
+            
+            var newCoins: [CoinMeta] = []
+            for asset in filteredSelection {
+                if !vault.coins.contains(where: { $0.ticker == asset.ticker && $0.chain == asset.chain}) {
+                    newCoins.append(asset)
+                    print("asset ticker \(asset.ticker)")
+                }
+            }
+            
+            try await addToChain(assets: newCoins, to: vault)
+            
+        } catch {
+            print("fail to save asset,\(error)")
+        }
+    }
+    
+    public func addToChain(assets: [CoinMeta], to vault: Vault) async throws {
+        if let coin = assets.first, coin.chain.chainType == .EVM, !coin.isNativeToken {
```

Comment:
this is very confusing , need to add some comments. 
inside the if , is add ERC20/BEP20 tokens
inside else is add chain

Reply:
I didn't change that. I only moved. I will separate the logic soon using a factory and some methods to instantiate the coins. So we can kill those if EVM, etc.